### ISSUE TITLE
Best practice when actions may be invalid in certain states

### ISSUE THREAD

**Issue by sigma-2** [joefarrington/mdpax#3]
Hi, again.

I've encountered an issue with the solver on the MDP problem I'm working. I'm not allowed to disclose the full information on the project, although I'll try to share as much relevant information and code to explain why I think something is going wrong.

States are defined as 4 integers in an array-like shape, i.e: [1, 0, 0, 0].

The odd positions are used to store partial results from previous actions, and is done so since I need to deal with multidimensional rewards. In this example, they are the ones in bold: [**1**, 0, **0**, 0]

The even positions are counters, that tell how many times the action related to the previous partial lefts can still be performed. [1, **0**, 0, **0**] in this case, no actions related to those two events can be performed.

I have three action, one related with each of the partial results, a last one that for any given state, leads to a final absorbing state. In this example, action 0 is related to the first partial result, and can be while its counter is strictly greater than 0. Similarly for action 1 and the second partial result. Lastly, action 2 is the one that leads to the final absorbing state.

```
Action space:
[[0]
 [1]
 [2]]
```

As for the events space. I have [-1, 1, -999]. -1 and 1 are intended to modify the partial sums, and -999 is related to action 2, and has no impact on actions 0 and 2.
```
Random event space
[[-1.00000e+00]
 [ 1.00000e+00]
 [-9.99999e+05]]
```

This is my probability matrix for this setup:
```
[[0.5 0.5 0. ]
 [0.5 0.5 0. ]
 [0.  0.  1. ]]
```

Although when actions are not available due to the counter, it will return 0:
```
state = jnp.array([0, 0, 1, 1])
action = jnp.array([0])
random_event = jnp.array([-1])
print("Random event probabilities:")
print(problem.random_event_probability(state=state, action=action,
                                               random_event=random_event))
```
Running this prints:
```
Random event probabilities:
0.0
```
As it should be, since that action is not available and returns 0 for any event.

---

Also, since that action is not allowed, when updating the state, it returns the same state:
```
print("Updated state")
print(problem.update_state_from_event(state=state, action=action,   random_event=random_event, verbose=1))
```
Prints:
```
Updated state
[0 0 1 1]
```
As expected.

---

Furthermore, when the action isn't valid, I code a penalty of -1e8 to ensure it shouldn't be optimal:
```
print(problem.transition(state=state,
                                 action=action,
                                 random_event=random_event))
```
Outputs:
```
(Array([0, 0, 1, 1], dtype=int32), Array(-1.e+08, dtype=float32))
```

---

Nevertheless, when runnint the solver, the optimal policy includes  states in which action 0 can't be performed and should be penalized as optimal, and with a reward of -0

```
print_solution(problem, solution)
```
outputs:
```
STATE               	POLICY	VALUE
0, 2, 0, 2	0	0.762500
0, 2, 0, 1	0	0.425000
0, 2, 1, 1	1	1.100000
0, 2, 0, 0	0	0.400000
0, 2, 1, 0	0	0.400000
0, 2, 2, 0	2	1.800000
0, 1, 0, 2	1	0.425000
0, 1, 0, 1	0	0.000000
0, 1, 1, 1	1	0.850000
0, 1, 0, 0	0	0.000000
0, 1, 1, 0	0	0.000000
0, 1, 2, 0	2	1.700000
1, 1, 0, 2	0	1.100000
1, 1, 0, 1	0	0.850000
1, 1, 1, 1	0	1.250000
1, 1, 0, 0	0	0.800000
1, 1, 1, 0	0	0.800000
1, 1, 2, 0	2	1.700000
0, 0, 0, 2	1	0.400000
0, 0, 0, 1	0	0.000000
0, 0, 1, 1	1	0.800000
0, 0, 0, 0	0	-0.000000
0, 0, 1, 0	0	-0.000000
0, 0, 2, 0	2	1.600000
1, 0, 0, 2	1	0.400000
1, 0, 0, 1	0	0.000000
1, 0, 1, 1	1	0.800000
1, 0, 0, 0	0	-0.000000
1, 0, 1, 0	0	-0.000000
1, 0, 2, 0	2	1.600000
2, 0, 0, 2	2	1.800000
2, 0, 0, 1	2	1.700000
2, 0, 1, 1	2	1.700000
2, 0, 0, 0	2	1.600000
2, 0, 1, 0	2	1.600000
2, 0, 2, 0	2	1.600000
-1, -1, -1, -1	0	0.000000
```
Notice these lines:
```
0, 0, 0, 1	0	0.000000
0, 0, 0, 0	0	-0.000000
0, 0, 1, 0	0	-0.000000
1, 0, 0, 1	0	0.000000
1, 0, 0, 0	0	-0.000000
1, 0, 1, 0	0	-0.000000
```
In which action 0 shouldn't be optimal, and is nevertheless the solution.
Also, it seems to happen only with action 0, I haven't seen cases in which this happens with action 1.

---

Any clues on how to address this?

Thank you for your help!

**Comment by joefarrington** [comment]
Hey, sorry I missed the notification for this. Did you get any further? 

Did you try both policy and value iteration? Would be interesting to know if they're giving the same/different optimal policies and might help narrow it down.

**Comment by sigma-2** [comment]
``` Solution Value Iteration
STATE               	POLICY	VALUE
0, 2, 0, 2	0	0.762500
0, 2, 0, 1	0	0.425000
0, 2, 1, 1	1	1.100000
0, 2, 0, 0	0	0.400000
0, 2, 1, 0	0	0.400000
0, 2, 2, 0	2	1.800000
0, 1, 0, 2	1	0.425000
0, 1, 0, 1	0	0.000000
0, 1, 1, 1	1	0.850000
0, 1, 0, 0	0	0.000000
0, 1, 1, 0	0	0.000000
0, 1, 2, 0	2	1.700000
1, 1, 0, 2	0	1.100000
1, 1, 0, 1	0	0.850000
1, 1, 1, 1	0	1.250000
1, 1, 0, 0	0	0.800000
1, 1, 1, 0	0	0.800000
1, 1, 2, 0	2	1.700000
0, 0, 0, 2	1	0.400000
0, 0, 0, 1	0	0.000000
0, 0, 1, 1	1	0.800000
0, 0, 0, 0	0	-0.000000
0, 0, 1, 0	0	-0.000000
0, 0, 2, 0	2	1.600000
1, 0, 0, 2	1	0.400000
1, 0, 0, 1	0	0.000000
1, 0, 1, 1	1	0.800000
1, 0, 0, 0	0	-0.000000
1, 0, 1, 0	0	-0.000000
1, 0, 2, 0	2	1.600000
2, 0, 0, 2	2	1.800000
2, 0, 0, 1	2	1.700000
2, 0, 1, 1	2	1.700000
2, 0, 0, 0	2	1.600000
2, 0, 1, 0	2	1.600000
2, 0, 2, 0	2	1.600000
-1, -1, -1, -1	0	0.000000
```

```Solution Policy iteration
STATE               	POLICY	VALUE
0, 2, 0, 2	0	0.762500
0, 2, 0, 1	0	0.425000
0, 2, 1, 1	1	1.100000
0, 2, 0, 0	0	0.400000
0, 2, 1, 0	0	0.400000
0, 2, 2, 0	2	1.800000
0, 1, 0, 2	1	0.425000
0, 1, 0, 1	0	0.000000
0, 1, 1, 1	1	0.850000
0, 1, 0, 0	0	0.000000
0, 1, 1, 0	0	0.000000
0, 1, 2, 0	2	1.700000
1, 1, 0, 2	0	1.100000
1, 1, 0, 1	0	0.850000
1, 1, 1, 1	0	1.250000
1, 1, 0, 0	0	0.800000
1, 1, 1, 0	0	0.800000
1, 1, 2, 0	2	1.700000
0, 0, 0, 2	1	0.400000
0, 0, 0, 1	0	-0.000000
0, 0, 1, 1	1	0.800000
0, 0, 0, 0	0	-0.000000
0, 0, 1, 0	0	-0.000000
0, 0, 2, 0	2	1.600000
1, 0, 0, 2	1	0.400000
1, 0, 0, 1	0	-0.000000
1, 0, 1, 1	1	0.800000
1, 0, 0, 0	0	-0.000000
1, 0, 1, 0	0	-0.000000
1, 0, 2, 0	2	1.600000
2, 0, 0, 2	2	1.800000
2, 0, 0, 1	2	1.700000
2, 0, 1, 1	2	1.700000
2, 0, 0, 0	2	1.600000
2, 0, 1, 0	2	1.600000
2, 0, 2, 0	2	1.600000
-1, -1, -1, -1	0	0.000000
```


The only differences are in states (0, 0, 0, 1) and (1, 0, 0, 1), lines 20 and 26, where the sign of the value changes from 0 to -0.

---


Invalid actions `0`
```Invalid Action 0
0, 0, 0, 1	0	0.000000
0, 0, 0, 0	0	-0.000000
0, 0, 1, 0	0	-0.000000
1, 0, 0, 1	0	0.000000
1, 0, 1, 1	1	0.800000
1, 0, 0, 0	0	-0.000000
1, 0, 1, 0	0	-0.000000
```

Invalid actions `1`
```Invalid
None
```


---

## Testing transitions and rewards.
I try the methods with these variables:
```Set up
state = jnp.array([0, 0, 0, 1]) # One of the state in the invalid actions
action = jnp.array([0]) # Invalid action
```

I try the 3 random events in this simple setup, and check the probability and updated state (and reward for it).

```
random_event = jnp.array([-1])
print("Random event probabilities:")
print(problem.random_event_probability(state=state, action=action,  
                                       random_event=random_event))
print("=============")  
print(problem.transition(state=state,  
                         action=action,  
                         random_event=random_event))
```

``` Print
Random event probabilities:
0.0
=============
(Array([0, 0, 0, 1], dtype=int32), Array(-1.e+08, dtype=float32))
```

It throws back the same state, and incurs in the penalty.

Then for event `[1]`
```
random_event = jnp.array([1])
```

```
Random event probabilities:
0.0
=============
(Array([0, 0, 0, 1], dtype=int32), Array(-1.e+08, dtype=float32))
```

It throws back the same state and incurs in the penalty

Lastly, event  `[-999999]`
```
random_event = jnp.array([-999999])
```

```
Random event probabilities:
0.0
=============
(Array([0, 0, 0, 1], dtype=int32), Array(-1.e+08, dtype=float32))
```

Same result.

---

I changed the code so that now the probability for the `[-999999]` event is 1.0 when the action isn't valid, and remains 0 for the other events

```
random_event = jnp.array([-999999])

# Prints:
Random event probabilities:
-1.0
=============
(Array([0, 0, 0, 1], dtype=int32), Array(-1.e+08, dtype=float32))
```


---

Running the solver with this tweak:

``` Value Iteration
STATE               	POLICY	VALUE
0, 2, 0, 2	0	0.500000
0, 2, 0, 1	0	0.000000
0, 2, 1, 1	0	1.000000
0, 2, 0, 0	2	-0.200000
0, 2, 1, 0	0	0.200000
0, 2, 2, 0	2	1.800000
0, 1, 0, 2	1	0.000000
0, 1, 0, 1	0	-0.800000
0, 1, 1, 1	2	0.800000
0, 1, 0, 0	2	-1.300000
0, 1, 1, 0	2	-0.300000
0, 1, 2, 0	2	1.700000
1, 1, 0, 2	0	1.000000
1, 1, 0, 1	2	0.800000
1, 1, 1, 1	0	1.200000
1, 1, 0, 0	2	0.700000
1, 1, 1, 0	2	0.700000
1, 1, 2, 0	2	1.700000
0, 0, 0, 2	2	-0.200000
0, 0, 0, 1	2	-1.300000
0, 0, 1, 1	2	0.700000
0, 0, 0, 0	2	-2.400000
0, 0, 1, 0	2	-0.400000
0, 0, 2, 0	2	1.600000
1, 0, 0, 2	1	0.200000
1, 0, 0, 1	2	-0.300000
1, 0, 1, 1	2	0.700000
1, 0, 0, 0	2	-0.400000
1, 0, 1, 0	2	-0.400000
1, 0, 2, 0	2	1.600000
2, 0, 0, 2	2	1.800000
2, 0, 0, 1	2	1.700000
2, 0, 1, 1	2	1.700000
2, 0, 0, 0	2	1.600000
2, 0, 1, 0	2	1.600000
2, 0, 2, 0	2	1.600000
-1, -1, -1, -1	0	0.000000
```

```Policy
0, 2, 0, 2	0	0.500000
0, 2, 0, 1	0	0.000000
0, 2, 1, 1	0	1.000000
0, 2, 0, 0	2	-0.200000
0, 2, 1, 0	0	0.200000
0, 2, 2, 0	2	1.800000
0, 1, 0, 2	1	0.000000
0, 1, 0, 1	0	-0.800000
0, 1, 1, 1	2	0.800000
0, 1, 0, 0	2	-1.300000
0, 1, 1, 0	2	-0.300000
0, 1, 2, 0	2	1.700000
1, 1, 0, 2	0	1.000000
1, 1, 0, 1	2	0.800000
1, 1, 1, 1	0	1.200000
1, 1, 0, 0	2	0.700000
1, 1, 1, 0	2	0.700000
1, 1, 2, 0	2	1.700000
0, 0, 0, 2	2	-0.200000
0, 0, 0, 1	2	-1.300000
0, 0, 1, 1	2	0.700000
0, 0, 0, 0	2	-2.400000
0, 0, 1, 0	2	-0.400000
0, 0, 2, 0	2	1.600000
1, 0, 0, 2	1	0.200000
1, 0, 0, 1	2	-0.300000
1, 0, 1, 1	2	0.700000
1, 0, 0, 0	2	-0.400000
1, 0, 1, 0	2	-0.400000
1, 0, 2, 0	2	1.600000
2, 0, 0, 2	2	1.800000
2, 0, 0, 1	2	1.700000
2, 0, 1, 1	2	1.700000
2, 0, 0, 0	2	1.600000
2, 0, 1, 0	2	1.600000
2, 0, 2, 0	2	1.600000
-1, -1, -1, -1	0	0.000000
```

Now both results are the same and do correspond with the exact solution. Indeed, the error was due to the 0 probability to all the events when the action wasn't allowed.


Beset and thank you for your help and pointers!
